server/protocols/jmap/src/main/java/org/apache/james/jmap/model/UpdateMessagePatch.java

```diff
@@ -125,18 +125,18 @@ public boolean isValid() {
         return getValidationErrors().isEmpty();
     }
 
-    public Flags applyToState(boolean isSeen, boolean isAnswered, boolean isFlagged) {
+    public Flags applyToState(Flags currentFlags) {
         Flags newStateFlags = new Flags();
 
-        boolean shouldMessageBeFlagged = isFlagged().isPresent() && isFlagged().get() || (!isFlagged().isPresent() && isFlagged);
+        boolean shouldMessageBeFlagged = isFlagged().isPresent() && isFlagged().get() || (!isFlagged().isPresent() && currentFlags.contains(Flags.Flag.FLAGGED));
```

Comment:
In fact it is equivalent to neither the first nor the second :)
Anyway I refactorized these ugly boolean expressions in pretty java 8 style.